### PR TITLE
Add support for block and mutex profiling

### PR DESCRIPTION
**Pull request by vickenty**
### What does this PR do?

- add runtime config settings for golang's runtime block and mutex profiling options
- make flare collect block and mutex profiles with every performance profile
- add flare options to enable block and mutex profiling before starting cpu profile
- pass runtime block and mutex profile settings to continuous profiler when starting internal profiling

### Motivation

Get better insights into blocking behavior of the agent.

### Describe how to test your changes

1. Collect a flare from a running agent using `agent flare -p 30 -B -M`
2. Check that the flare archive contains `core-block.pprof` and `core-mutex.pprof`
3. At least block profile should have some data: `go tool pprof -top core-block.pprof`; mutex might end up empty on a lightly loaded agent
4. Restart the agent
5. Set runtime settings
   ```
   agent config set runtime_block_profile_rate 1
   agent config set runtime_mutex_profile_fraction 1
   ```
7. Collect a flare using `agent flare -p 30`
7. Flare archive should contain `core-block.pprof` and `core-mutex.pprof`.
8. Both files should have content, check using `go tool pprof -top <file>`
9. Collect a flare without access to auth token: `agent flare -p 30` (either use a non-root user or `chmod 0 /etc/datadog-agent/auth_token`)
10. Flare command should complete without errors by falling back to local mode.
9. Enable internal profiling for 10 minutes: `agent config set internal_profiling true`
10. Verify that cpu, heap, block and mutex profiles are visible in the profiling section of the app (find your host and click on the profile under the graphs)
12. Enable internal profiling in the configuration file:
    ```
    internal_profiling:
      enabled: true
      block_profile_rate 1
      mutex_profile_fraction: 1
      enable_goroutine_stacktraces: true
    ```
13. Restart the agent
14. After 10 minutes verify that cpu, heap, block, mutex and goroutine profiles are visible in the profiling section of the app (find your host and click on the profile under the graphs)

### CODE REVIEW

File: cmd/agent/app/flare.go [hunk 1]
```diff
@@ -79,6 +92,16 @@ var flareCmd = &cobra.Command{
 }
 
 func readProfileData(pdata *flare.ProfileData) error {
+	c, err := common.NewSettingsClient()
+	if err != nil {
+		return fmt.Errorf("failed to initialize settings client: %v", err)
+	}
+	prevSettings := make(map[string]interface{})
+	if err := setRuntimeProfilingSettings(c, &prevSettings); err != nil {
```
Comment (ogaca-dd) [inline cmd/agent/app/flare.go:100]: Did you consider passing `runtime_block_profile_rate` and `runtime_block_profile_rate` as parameters to the api call like `profiling` (The api handler can save the previous state and restore it when done) 
One of the advantage is that you can curl the url `http://127.0.0.1:%s/debug/pprof/...` without manually restore the profile rates.

There is no need to change anything, I am just curious if it would be possible.

File: cmd/agent/app/run.go [hunk 2]
```diff
@@ -253,6 +253,16 @@ func StartAgent() error {
 	}
 
 	// Setup Internal Profiling
+	if v := config.Datadog.GetInt("internal_profiling.block_profile_rate"); v > 0 {
```
Comment (ogaca-dd) [inline cmd/agent/app/run.go:256]: I am not sure we need a configuration option from `datadog.yaml`. Do you have use case in mind?

File: pkg/config/settings/runtime_setting_block_profile_rate.go [hunk 3]
```diff
@@ -35,6 +36,7 @@ func (r RuntimeBlockProfileRate) Set(value interface{}) error {
 	}
 
 	profiling.SetBlockProfileRate(rate)
+	config.Datadog.Set("internal_profiling.block_profile_rate", rate)
```
Comment (ogaca-dd) [inline pkg/config/settings/runtime_setting_block_profile_rate.go:39]: What is the reason to change the config? When you set something in `datadog.yaml` and later call `config.Datadog.Get("Something")` you expect to have the value from the config file.

File: cmd/agent/app/flare.go [hunk 4]
```diff
@@ -53,6 +62,10 @@ var flareCmd = &cobra.Command{
 			return fmt.Errorf("unable to set up global agent configuration: %v", err)
 		}
 
+		if err := util.SetAuthToken(); err != nil {
```
Comment (ogaca-dd) [inline cmd/agent/app/flare.go:65]: What is the reason for this change?

File: pkg/config/settings/runtime_setting_mutex_profile_fraction.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+// Set changes the value of the runtime setting
+func (r RuntimeMutexProfileFraction) Set(value interface{}) error {
+	rate, err := GetInt(value)
+	if err != nil {
+		return err
+	}
+
+	profiling.SetMutexProfileFraction(rate)
+	config.Datadog.Set("internal_profiling.mutex_profile_fraction", rate)
```
Comment (ogaca-dd) [inline pkg/config/settings/runtime_setting_mutex_profile_fraction.go:39]: What is the reason to change the config?

File: pkg/flare/archive.go [hunk 6]
```diff
@@ -113,6 +113,16 @@ func CreatePerformanceProfile(prefix, debugURL string, cpusec int, target *Profi
 			Name: prefix + "-2nd-heap.pprof",
 			URL:  debugURL + "/heap",
 		},
+		{
+			// mutex profile
+			Name: prefix + "-mutex.pprof",
+			URL:  debugURL + "/mutex",
```
Comment (ogaca-dd) [inline pkg/flare/archive.go:119]: Where this endpoint is defined?

File: pkg/flare/archive.go [hunk 7]
```diff
@@ -113,6 +113,16 @@ func CreatePerformanceProfile(prefix, debugURL string, cpusec int, target *Profi
...
+		},
+		{
+			// goroutine blocking profile
+			Name: prefix + "-block.pprof",
+			URL:  debugURL + "/block",
```
Comment (ogaca-dd) [inline pkg/flare/archive.go:124]: Where this endpoint is defined?

**Review by ogaca-dd — COMMENTED** [review]
Nice PR. 
I have added several comments but minor ones.

File: pkg/flare/archive.go (hunk 7 above)
Comment (vickenty) [reply pkg/flare/archive.go:124]: It is provided by `pprof` package: https://golang.org/pkg/net/http/pprof/

File: pkg/flare/archive.go (hunk 6 above)
Comment (vickenty) [reply pkg/flare/archive.go:119]: It is provided by `pprof` package: https://golang.org/pkg/net/http/pprof/

File: pkg/config/settings/runtime_setting_mutex_profile_fraction.go (hunk 5 above)
Comment (vickenty) [reply pkg/config/settings/runtime_setting_mutex_profile_fraction.go:39]: So that the current values are reflected in the runtime config dump returned by `agent config`

File: pkg/config/settings/runtime_setting_block_profile_rate.go (hunk 3 above)
Comment (vickenty) [reply pkg/config/settings/runtime_setting_block_profile_rate.go:39]: So that the current values are reflected in the runtime config dump returned by `agent config`. Also for consistency with other runtime settings, which do the same thing.

File: cmd/agent/app/run.go (hunk 2 above)
Comment (vickenty) [reply cmd/agent/app/run.go:256]: If we enable internal profiling in the config, we may want to configure these optional profile types at the same time.

File: cmd/agent/app/flare.go (hunk 1 above)
Comment (vickenty) [reply cmd/agent/app/flare.go:100]: Entire `/debug/pprof` is provided as-is by the pprof package, including the `profile` endpoint we call to do the profiling. We could override it, of course, that would be bit surprising. I don't think a curl user would expect the endpoint to have additional undocumented parameters.

File: pkg/config/settings/runtime_setting.go [hunk 8]
```diff
@@ -95,3 +96,22 @@ func GetBool(v interface{}) (bool, error) {
...
+// GetInt returns the integer value contained in value.
+// If value is a integer, returns its value
+// If value is a string, it parses the string into an integer.
+// Else, returns an error.
+func GetInt(v interface{}) (int, error) {
+	switch v := v.(type) {
+	case int:
+		return v, nil
+	case string:
+		i, err := strconv.ParseInt(v, 10, 32)
```
[the remark this answers is not in the document]
Comment (vickenty) [reply pkg/config/settings/runtime_setting.go]: The idea was to use the smaller size to avoid overflow in `int(i)` later, but `ParseInt` accepts `bitWidth=0` to indicate `int` size, so that's even better.

File: cmd/agent/app/flare.go (hunk 4 above)
Comment (vickenty) [reply cmd/agent/app/flare.go:65]: This is necessary to call agent api to change runtime settings.

File: cmd/agent/app/flare.go [hunk 9]
```diff
@@ -207,3 +231,46 @@ func createArchive(logFiles []string, pdata flare.ProfileData) (string, error) {
...
+	}
+
+	return oldVal, nil
+}
+
+func resetRuntimeProfilingSettings(c settings.Client, prev map[string]interface{}) {
+	for name, value := range prev {
+		fmt.Fprintln(color.Output, color.BlueString("Restoring %s to %v", name, value))
+		if _, err := c.Set(name, fmt.Sprint(value)); err != nil {
+			fmt.Fprintln(color.Output, color.RedString("Failed restore previous value of %s: %v", name, err))
```
Comment (olivielpeau) [inline cmd/agent/app/flare.go:273]:
```suggestion
			fmt.Fprintln(color.Output, color.RedString("Failed to restore previous value of %s: %v", name, err))
```

File: pkg/config/settings/runtime_setting_block_profile_rate.go [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+type RuntimeBlockProfileRate (string)
+
+// Name returns the name of the runtime setting
+func (r RuntimeBlockProfileRate) Name() string {
+	return string(r)
+}
+
+// Description returns the runtime setting's description
+func (r RuntimeBlockProfileRate) Description() string {
+	return "This setting controls the fraction of goroutine blocking events that are reported in the blocking profile"
```
Comment (olivielpeau) [inline pkg/config/settings/runtime_setting_block_profile_rate.go:18]: Let's mention explicitly that this only affects _internal_  profiles (not application profiles that the trace-agent forwards to DD)

```suggestion
	return "This setting controls the fraction of goroutine blocking events that are reported in the internal blocking profile"
```

File: pkg/config/settings/runtime_setting_mutex_profile_fraction.go [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+type RuntimeMutexProfileFraction (string)
+
+// Name returns the name of the runtime setting
+func (r RuntimeMutexProfileFraction) Name() string {
+	return string(r)
+}
+
+// Description returns the runtime setting's description
+func (r RuntimeMutexProfileFraction) Description() string {
+	return "This setting controls the fraction of mutex contention events that are reported in the mutex profile"
```
Comment (olivielpeau) [inline pkg/config/settings/runtime_setting_mutex_profile_fraction.go:18]:
```suggestion
	return "This setting controls the fraction of mutex contention events that are reported in the internal mutex profile"
```

File: pkg/util/profiling/profiling.go [hunk 12]
```diff
@@ -48,7 +49,20 @@ func Start(apiKey, site, env, service string, period time.Duration, cpuDuration
 		profiler.WithProfileTypes(profiler.CPUProfile, profiler.HeapProfile, profiler.MutexProfile),
```
Comment (olivielpeau) [inline pkg/util/profiling/profiling.go:49]: Do we need to also enable the relevant profile types here to actually enable them in the DD profiler? https://docs.datadoghq.com/tracing/profiler/enabling/#pagetitle

File: pkg/util/profiling/runtime.go [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+	mu.RLock()
+	defer mu.RUnlock()
+	return blockProfileRate
+}
+
+// SetMutexProfileFraction sets fraction of mutexes that generate profiling data.
+func SetMutexProfileFraction(fraction int) {
+	mu.Lock()
+	defer mu.Unlock()
+	runtime.SetMutexProfileFraction(fraction)
```
Comment (olivielpeau) [inline pkg/util/profiling/runtime.go:33]: Does this call only set a profiling option on the Go runtime, or does it also make the Go runtime start profiling?

(same question about `runtime.SetBlockProfileRate`)

**Review by olivielpeau — COMMENTED** [review]
As part of the testing and QA of this change, we should also validate that when the relevant `internal_profiling` options are enabled (through runtime settings, or through static agent config), the related profiles are visible in-app in DD profiler (https://app.datadoghq.com/profiling)

File: pkg/util/profiling/profiling.go (hunk 12 above)
Comment (olivielpeau) [reply pkg/util/profiling/profiling.go:49]: additional question for @Gui774ume : `profiler.MutexProfile` was added by default in https://github.com/DataDog/datadog-agent/pull/8025, but since it's documented as having additional overhead, wouldn't it be better to only enable it if an additional config option is enabled? (which is the approach of the present PR for the core agent)

Not urgent for 7.29 since internal profiling is an internal feature, but for 7.30.

File: cmd/agent/common/common.go [hunk 14]
```diff
@@ -68,3 +72,13 @@ func GetVersion(w http.ResponseWriter, r *http.Request) {
...
 }
+
+// NewSettingsClient returns a configured runtime settings client.
+func NewSettingsClient() (settings.Client, error) {
+	ipcAddress, err := config.GetIPCAddress()
+	if err != nil {
+		return nil, err
+	}
+	hc := util.GetClient(false)
+	return settingshttp.NewClient(hc, fmt.Sprintf("https://%v:%v/agent/config", ipcAddress, config.Datadog.GetInt("cmd_port")), "datadog-agent"), nil
```
Comment (olivielpeau) [inline cmd/agent/common/common.go:83]: nit: we generally refer to the core agent process as `agent` rather than `datadog-agent`

```suggestion
	return settingshttp.NewClient(hc, fmt.Sprintf("https://%v:%v/agent/config", ipcAddress, config.Datadog.GetInt("cmd_port")), "agent"), nil
```

File: pkg/util/profiling/runtime.go (hunk 13 above)
Comment (vickenty) [reply pkg/util/profiling/runtime.go:33]: Yes, setting these to non-zero values makes go runtime to start collecting profiling data.

**Comment by olivielpeau** [comment]
Let's update the profiler (`dd-trace-go`) to the latest version as well while we're at it

File: cmd/agent/app/flare.go (hunk 4 above)
Comment (ogaca-dd) [reply cmd/agent/app/flare.go:65]: A user without an access to the auth token can now run `agent flare`, with this change it will not be possible. I am not sure this is what we want.

File: pkg/util/profiling/runtime.go (hunk 13 above)
Comment (olivielpeau) [reply pkg/util/profiling/runtime.go:33]: ok, is there any risk to leaving them on "forever"? (does it use more and more memory over time for example?). If so, I think these runtime settings should be hidden (and worth a comment in the code explaining why they're hidden)

File: pkg/util/profiling/profiling.go (hunk 12 above)
Comment (vickenty) [reply pkg/util/profiling/profiling.go:49]: > Do we need to also enable the relevant profile types here to actually enable them in the DD profiler?
No, they will be enabled by passing `BlockProfileRate` and `MutexProfileFraction` options.

File: pkg/util/profiling/runtime.go (hunk 13 above)
Comment (vickenty) [reply pkg/util/profiling/runtime.go:33]: These features do allocate a small amount per-stack frame sampled. I would expect allocations to taper off with time, as the profiler accumulates location data; but there is risk that number of possible stack frames to be too large. This would cause not only increased memory usage, but also additional cpu cost (profiling data is stored in a simple fixed-size linked-list hashmap with a fairly simple hash function). Since this is a concern, I will make these settings hidden.

File: pkg/config/settings/runtine_setting_profiling_goroutines.go [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+// Set changes the value of the runtime setting
+func (r ProfilingGoroutines) Set(value interface{}) error {
+	enabled, err := GetBool(value)
+	if err != nil {
+		return err
+	}
+
+	config.Datadog.Set("internal_profiling.enable_goroutine_stacktraces", enabled)
```
Comment (ogaca-dd) [inline pkg/config/settings/runtine_setting_profiling_goroutines.go:37]: Is `config.Datadog.Set` thread safe?

Comment (vickenty) [reply pkg/config/settings/runtine_setting_profiling_goroutines.go:37]: [Yes](https://github.com/DataDog/datadog-agent/blob/971434bcef8ff208132d2b723ed63923bc907cd6/pkg/config/viper.go#L34)

**Review by brycekahle — APPROVED** [review]

**Review by olivielpeau — APPROVED** [review]
This looks great, thanks ✨

**Review by xornivore — APPROVED** [review]
Awesome 🚀

File: cmd/agent/app/flare.go (hunk 4 above)
Comment (vickenty) [reply cmd/agent/app/flare.go]: https://github.com/DataDog/datadog-agent/pull/8360/commits/e165515386ddef90398c5d1545ca9c3acc49a3e1

**Review by ogaca-dd — APPROVED** [review]
